#### Lisa Over
CMPINF 2110<br>
Homework 2<br>
February 8, 2021

In [1]:
import sys
print(sys.executable)

/Users/lisa/opt/anaconda3/envs/cmpinf_2110/bin/python


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import seaborn as sns

## Problem 01:  Tidy the Tuberculosis Data

#### Read and display the data

In [3]:
file_name = "tidy_data_table_9_use.csv"
tb_df = pd.read_csv(file_name)

In [4]:
tb_df.head(8)

country  year  f_014  f_1524  f_2534  f_3544  f_4554  f_5564  f_65  f_u  \
0      AD  1989    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN   
1      AD  1990    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN   
2      AD  1991    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN   
3      AD  1992    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN   
4      AD  1993    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN   
5      AD  1994    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN   
6      AD  1996    0.0     1.0     1.0     0.0     0.0     1.0   0.0  NaN   
7      AD  1997    0.0     1.0     2.0     3.0     0.0     0.0   1.0  NaN   

   m_014  m_1524  m_2534  m_3544  m_4554  m_5564  m_65  m_u  
0    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN  
1    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN  
2    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN  
3    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN  
4    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN  
5    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN  
6    0.0     0.0     0.0     4.0     1.0     0.0   0.0  NaN  
7    0.0     0.0     1.0     2.0     2.0     1.0   6.0  NaN

In [5]:
tb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5769 entries, 0 to 5768
Data columns (total 18 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   country  5746 non-null   object 
 1   year     5769 non-null   int64  
 2   f_014    2378 non-null   float64
 3   f_1524   2394 non-null   float64
 4   f_2534   2400 non-null   float64
 5   f_3544   2398 non-null   float64
 6   f_4554   2401 non-null   float64
 7   f_5564   2394 non-null   float64
 8   f_65     2396 non-null   float64
 9   f_u      272 non-null    float64
 10  m_014    2381 non-null   float64
 11  m_1524   2407 non-null   float64
 12  m_2534   2408 non-null   float64
 13  m_3544   2415 non-null   float64
 14  m_4554   2421 non-null   float64
 15  m_5564   2414 non-null   float64
 16  m_65     2408 non-null   float64
 17  m_u      273 non-null    float64
dtypes: float64(16), int64(1), object(1)
memory usage: 811.4+ KB


In [6]:
tb_df.country.value_counts()

UY    29
AU    29
NO    29
BJ    29
TT    29
      ..
AN    13
VG    10
TL     8
ME     4
RS     4
Name: country, Length: 212, dtype: int64

#### Stack the columns

The tb_df dataset has headers as values instead on all headers being variable names. Except for 'country' and 'year,' the headers contain sex and age information. The 'm' or 'f' at the beginning indicate male or female, respectively. The numbers indicate age group, for example, '014' means the age group '0-14.'

Save the variable names except, for 'country' and 'year' and stack the columns with melt.

In [7]:
var_names = tb_df.columns.to_list()
var_names

['country',
 'year',
 'f_014',
 'f_1524',
 'f_2534',
 'f_3544',
 'f_4554',
 'f_5564',
 'f_65',
 'f_u',
 'm_014',
 'm_1524',
 'm_2534',
 'm_3544',
 'm_4554',
 'm_5564',
 'm_65',
 'm_u']

In [8]:
var_names.remove('year')
var_names.remove('country')
var_names

['f_014',
 'f_1524',
 'f_2534',
 'f_3544',
 'f_4554',
 'f_5564',
 'f_65',
 'f_u',
 'm_014',
 'm_1524',
 'm_2534',
 'm_3544',
 'm_4554',
 'm_5564',
 'm_65',
 'm_u']

In [9]:
tb_stack = tb_df.reset_index().melt( id_vars = ['country', 'year'], value_vars = var_names )

In [10]:
tb_stack.head()

country  year variable  value
0      AD  1989    f_014    NaN
1      AD  1990    f_014    NaN
2      AD  1991    f_014    NaN
3      AD  1992    f_014    NaN
4      AD  1993    f_014    NaN

#### Split the new column into two

The new 'variable' column represents multiple variables: sex and gender. Split this column into two new columns and drop the 'variable' column.

In [11]:
tb_stack.variable.str.split( '_', expand=True)

0    1
0      f  014
1      f  014
2      f  014
3      f  014
4      f  014
...   ..  ...
92299  m    u
92300  m    u
92301  m    u
92302  m    u
92303  m    u

[92304 rows x 2 columns]

In [12]:
tb_stack[['sex','age']] = tb_stack.variable.str.split( '_', expand=True)

In [13]:
tb_stack.head()

country  year variable  value sex  age
0      AD  1989    f_014    NaN   f  014
1      AD  1990    f_014    NaN   f  014
2      AD  1991    f_014    NaN   f  014
3      AD  1992    f_014    NaN   f  014
4      AD  1993    f_014    NaN   f  014

In [14]:
tb_stack.drop(['variable'], axis=1, inplace=True)

In [15]:
tb_stack.head()

country  year  value sex  age
0      AD  1989    NaN   f  014
1      AD  1990    NaN   f  014
2      AD  1991    NaN   f  014
3      AD  1992    NaN   f  014
4      AD  1993    NaN   f  014

#### Convert age_group values

Check data types, list values and counts for age_group variable, create dictionary for age group lookup, and convert age values to new age_group column per dictionary entries.

In [16]:
tb_stack.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 92304 entries, 0 to 92303
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   country  91936 non-null  object 
 1   year     92304 non-null  int64  
 2   value    34160 non-null  float64
 3   sex      92304 non-null  object 
 4   age      92304 non-null  object 
dtypes: float64(1), int64(1), object(3)
memory usage: 3.5+ MB


In [17]:
tb_stack.age.value_counts()

3544    11538
5564    11538
65      11538
2534    11538
1524    11538
014     11538
4554    11538
u       11538
Name: age, dtype: int64

Make the dictionary for age group lookup. Set value for "u" equal to "NA" per cleaning code for TB data on github: https://github.com/hadley/tidy-data/blob/master/data/tb.r

In [18]:
age_group_dict = {"014":"0-14", "1524":"15-24", "2534":"25-34", "3544":"35-44", "4554":"45-54", "5564":"55-64", "65":"65+", "u":"NA"}

In [19]:
def convert_x(x, dict):
    return dict[x]

In [20]:
tb_stack['age_group'] = tb_stack.apply(lambda x: convert_x(x['age'],age_group_dict),axis=1)

In [21]:
tb_stack.head()

country  year  value sex  age age_group
0      AD  1989    NaN   f  014      0-14
1      AD  1990    NaN   f  014      0-14
2      AD  1991    NaN   f  014      0-14
3      AD  1992    NaN   f  014      0-14
4      AD  1993    NaN   f  014      0-14

In [22]:
tb_stack.drop(['age'], axis=1, inplace=True)

In [23]:
tb_stack.head()

country  year  value sex age_group
0      AD  1989    NaN   f      0-14
1      AD  1990    NaN   f      0-14
2      AD  1991    NaN   f      0-14
3      AD  1992    NaN   f      0-14
4      AD  1993    NaN   f      0-14

In [24]:
tb_tidy = tb_stack.copy()

#### Discussion on the tidy Tuberculosis data

There are 92,304 rows and 5 columns: country, year, value, sex, age_group. There are 212 unique values for country for both the original dataset and the tidy dataset. There are 29 unique year values. The number of entries for each year is displayed below.

In [25]:
tb_tidy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 92304 entries, 0 to 92303
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   country    91936 non-null  object 
 1   year       92304 non-null  int64  
 2   value      34160 non-null  float64
 3   sex        92304 non-null  object 
 4   age_group  92304 non-null  object 
dtypes: float64(1), int64(1), object(3)
memory usage: 3.5+ MB


In [26]:
tb_df.country.value_counts()

UY    29
AU    29
NO    29
BJ    29
TT    29
      ..
AN    13
VG    10
TL     8
ME     4
RS     4
Name: country, Length: 212, dtype: int64

In [27]:
tb_tidy.country.value_counts()

IN    464
KG    464
KH    464
CA    464
CY    464
     ... 
AN    208
VG    160
TL    128
ME     64
RS     64
Name: country, Length: 212, dtype: int64

In [28]:
tb_tidy.year.value_counts()

2008    3392
2006    3392
2007    3392
2002    3328
2003    3328
2004    3312
2005    3296
2001    3280
1998    3248
1988    3216
1999    3216
2000    3216
1996    3200
1987    3184
1997    3168
1985    3168
1989    3152
1986    3152
1995    3152
1983    3136
1990    3136
1981    3104
1982    3104
1984    3088
1991    3072
1980    3056
1992    2992
1994    2944
1993    2880
Name: year, dtype: int64

In [29]:
tb_tidy.year.nunique()

29

In [79]:
tb_tidy.to_csv('tb_tidy.csv')

## Problem 02: Group and Aggregate the Tuberculosis Data

#### Which year has the most missing values? Which year has the fewest missing values?

The year 1988 has the most missing values with 3,202 missing. The year 2008 has the fewest missing values with 485 missing.

In [30]:
missing_count = pd.DataFrame(tb_tidy.groupby(["year"])["value"].size().to_frame('value') - tb_tidy.groupby(["year"])["value"].count().to_frame('value'))
missing_count.sort_values(by=["value"])

value
year       
2008    485
2006    640
2005    658
2004    666
2003    667
2007    775
2002    812
2000    943
1998   1011
2001   1033
1999   1052
1997   1054
1996   1100
1995   1284
1993   2852
1994   2888
1992   2964
1980   3028
1991   3044
1984   3060
1982   3076
1981   3076
1983   3108
1990   3122
1989   3124
1986   3124
1985   3140
1987   3156
1988   3202

In [107]:
missing_df = tb_tidy.loc[tb_tidy['value'].isna(), :]
missing_df.head()

country  year  value sex age_group
0      AD  1989    NaN   f      0-14
1      AD  1990    NaN   f      0-14
2      AD  1991    NaN   f      0-14
3      AD  1992    NaN   f      0-14
4      AD  1993    NaN   f      0-14

In [32]:
missing_df.groupby(["year"]).size().sort_values(ascending=False)

year
1988    3202
1987    3156
1985    3140
1986    3124
1989    3124
1990    3122
1983    3108
1981    3076
1982    3076
1984    3060
1991    3044
1980    3028
1992    2964
1994    2888
1993    2852
1995    1284
1996    1100
1997    1054
1999    1052
2001    1033
1998    1011
2000     943
2002     812
2007     775
2003     667
2004     666
2005     658
2006     640
2008     485
dtype: int64

#### Which year has the most total cases, across all age groups, for males in the countries with codes 'AO' and 'AR'?

* The year 2008 has the most cases (11,881 total cases) for males in country with code 'AO'.
* The year 2001 has the most cases (3,177 total cases) for males in country with code 'AR'.

In [33]:
tb_tidy.loc[tb_tidy['country']=='AO', :]

country  year  value sex age_group
187        AO  1980    NaN   f      0-14
188        AO  1981    NaN   f      0-14
189        AO  1982    NaN   f      0-14
190        AO  1983    NaN   f      0-14
191        AO  1984    NaN   f      0-14
...       ...   ...    ...  ..       ...
86746      AO  2004    NaN   m        NA
86747      AO  2005    NaN   m        NA
86748      AO  2006    NaN   m        NA
86749      AO  2007    NaN   m        NA
86750      AO  2008    0.0   m        NA

[464 rows x 5 columns]

In [34]:
tb_tidy.loc[(tb_tidy['country']=='AO') & (tb_tidy['sex']=='m'), :].groupby(['year']).\
aggregate(total_cases = ('value', 'sum')).\
reset_index().sort_values(by=['total_cases'], ascending=False)

year  total_cases
28  2008      11881.0
27  2007      11281.0
26  2006      10957.0
25  2005      10165.0
24  2004       9973.0
23  2003       9233.0
22  2002       9120.0
19  1999       4801.0
16  1996       4259.0
17  1997       4181.0
20  2000       4088.0
18  1998       3768.0
21  2001       3312.0
15  1995       2411.0
0   1980          0.0
1   1981          0.0
13  1993          0.0
12  1992          0.0
11  1991          0.0
10  1990          0.0
9   1989          0.0
8   1988          0.0
7   1987          0.0
6   1986          0.0
5   1985          0.0
4   1984          0.0
3   1983          0.0
2   1982          0.0
14  1994          0.0

In [35]:
tb_tidy.loc[(tb_tidy['country']=='AR') & (tb_tidy['sex']=='m'), :].groupby(['year']).\
aggregate(total_cases = ('value', 'sum')).\
reset_index().sort_values(by=['total_cases'], ascending=False)

year  total_cases
21  2001       3177.0
17  1997       3098.0
22  2002       3068.0
18  1998       2929.0
27  2007       2885.0
23  2003       2873.0
28  2008       2808.0
19  1999       2789.0
24  2004       2658.0
25  2005       2645.0
20  2000       2488.0
26  2006       2448.0
16  1996       2389.0
15  1995          0.0
0   1980          0.0
1   1981          0.0
13  1993          0.0
12  1992          0.0
11  1991          0.0
10  1990          0.0
9   1989          0.0
8   1988          0.0
7   1987          0.0
6   1986          0.0
5   1985          0.0
4   1984          0.0
3   1983          0.0
2   1982          0.0
14  1994          0.0

#### Which year has the most total cases, across age groups, for females in the countries with codes 'AF' and 'AM'?

* The year 2007 has the most cases (9,031 total cases) for females in the country with codes 'AF'.
* The year 2008 has the most cases (118 total cases) for females in the country with codes 'AM'.

In [36]:
tb_tidy.loc[(tb_tidy['country']=='AF') & (tb_tidy['sex']=='f'), :].groupby(['year']).\
aggregate(total_cases = ('value', 'sum')).\
reset_index().sort_values(by=['total_cases'], ascending=False)

year  total_cases
22  2007       9031.0
23  2008       8915.0
21  2006       8520.0
20  2005       6818.0
19  2004       5587.0
18  2003       4423.0
17  2002       4418.0
16  2001       3062.0
15  2000       1751.0
13  1998       1207.0
14  1999        517.0
12  1997        102.0
1   1981          0.0
11  1991          0.0
10  1990          0.0
9   1989          0.0
8   1988          0.0
7   1987          0.0
6   1986          0.0
5   1985          0.0
4   1984          0.0
3   1983          0.0
2   1982          0.0
0   1980          0.0

In [37]:
tb_tidy.loc[(tb_tidy['country']=='AM') & (tb_tidy['sex']=='f'), :].groupby(['year']).\
aggregate(total_cases = ('value', 'sum')).\
reset_index().sort_values(by=['total_cases'], ascending=False)

year  total_cases
28  2008        118.0
21  2001        116.0
22  2002        115.0
26  2006        102.0
23  2003         99.0
20  2000         96.0
27  2007         87.0
19  1999         85.0
25  2005         83.0
17  1997         80.0
18  1998         74.0
24  2004         70.0
16  1996         41.0
15  1995         13.0
0   1980          0.0
1   1981          0.0
13  1993          0.0
12  1992          0.0
11  1991          0.0
10  1990          0.0
9   1989          0.0
8   1988          0.0
7   1987          0.0
6   1986          0.0
5   1985          0.0
4   1984          0.0
3   1983          0.0
2   1982          0.0
14  1994          0.0

## Problem 03: Tidy the Stock Market Data

#### Read and display the data from each file

In [80]:
file_name = "daily_open_prices.csv"
open_df = pd.read_csv(file_name)
open_df.head()

date   AAPL_open    AMZN_open  GME_open    KO_open   TSLA_open  \
0  2020-09-01  132.759995  3489.580078      7.30  49.310001  502.140015   
1  2020-09-02  137.589996  3547.000000      7.80  49.360001  478.989990   
2  2020-09-03  126.910004  3485.000000      7.88  51.400002  407.230011   
3  2020-09-04  120.070000  3318.000000      7.78  50.939999  402.809998   
4  2020-09-08  113.949997  3144.000000      7.55  50.930000  356.000000   

     VZ_open  
0  59.119999  
1  59.150002  
2  60.830002  
3  60.480000  
4  60.799999

In [81]:
file_name = "daily_close_prices.csv"
close_df = pd.read_csv(file_name)
close_df.head()

date  AAPL_close   AMZN_close  GME_close   KO_close  TSLA_close  \
0  2020-09-01  134.179993  3499.120117       7.65  49.119999  475.049988   
1  2020-09-02  131.399994  3531.449951       7.71  51.189999  447.369995   
2  2020-09-03  120.879997  3368.000000       7.82  50.480000  407.000000   
3  2020-09-04  120.959999  3294.620117       7.65  51.040001  418.320007   
4  2020-09-08  112.820000  3149.840088       7.70  49.810001  330.209991   

    VZ_close  
0  59.160000  
1  60.529999  
2  60.610001  
3  60.480000  
4  59.959999

In [82]:
file_name = "daily_low_prices.csv"
low_df = pd.read_csv(file_name)
low_df.head()

date    AAPL_low     AMZN_low  GME_low     KO_low    TSLA_low  \
0  2020-09-01  130.529999  3467.000000     6.77  48.830002  470.510010   
1  2020-09-02  127.000000  3486.689941     7.11  49.360001  405.119995   
2  2020-09-03  120.500000  3303.000000     7.24  50.099998  402.000000   
3  2020-09-04  110.889999  3111.129883     7.17  50.279999  372.019989   
4  2020-09-08  112.680000  3130.000000     7.48  49.360001  329.880005   

      VZ_low  
0  58.779999  
1  59.000000  
2  60.139999  
3  60.060001  
4  59.619999

In [83]:
file_name = "daily_high_prices.csv"
high_df = pd.read_csv(file_name)
high_df.head()

date   AAPL_high    AMZN_high  GME_high    KO_high   TSLA_high  \
0  2020-09-01  134.800003  3513.870117      7.82  49.349998  502.489990   
1  2020-09-02  137.979996  3552.250000      8.05  51.360001  479.040009   
2  2020-09-03  128.839996  3488.409912      8.45  52.139999  431.799988   
3  2020-09-04  123.699997  3381.500000      7.92  51.470001  428.000000   
4  2020-09-08  118.989998  3250.850098      8.28  51.040001  368.739990   

     VZ_high  
0  59.299999  
1  60.740002  
2  61.279999  
3  61.500000  
4  60.799999

In [84]:
file_name = "stock_info.csv"
info_df = pd.read_csv(file_name)
info_df

symbol                      company                  sector exchange
0   AAPL                   Apple Inc.  Information Technology   NASDAQ
1     VZ  Verizon Communications Inc.  Communication Services     NYSE
2     KO            Coca-Cola Company        Consumer Staples     NYSE
3   AMZN              Amazon.com Inc.  Consumer Discretionary   NASDAQ
4   TSLA                    Tesla Inc  Consumer Discretionary   NASDAQ
5    GME               GameStop Corp.  Consumer Discretionary     NYSE

#### Stack columns of each price dataset

The four stock prices datasets - open_df, close_df, low_df, and high_df - each have headers that are values not variables. The first set of characters before the "_" are acronyms or symbols that indicate the company, sector, and exchange per the info_df dataset. The second set of characters - open, close, low, and high - describe the price as being the opening, closing, low, or high price, respectively, for a particular stock company, sector, and exchange on a particular day. 

Create list of variable names and then use melt to stack the columns of each stock prices dataset so that the values in the header become part of the observational unit of each row.

In [85]:
var_names = open_df.columns.to_list()
var_names.remove('date')
print(var_names)

['AAPL_open', 'AMZN_open', 'GME_open', 'KO_open', 'TSLA_open', 'VZ_open']


In [86]:
open_stack = open_df.reset_index().melt( id_vars = ['date'], value_vars = var_names )
open_stack.head()

date   variable       value
0  2020-09-01  AAPL_open  132.759995
1  2020-09-02  AAPL_open  137.589996
2  2020-09-03  AAPL_open  126.910004
3  2020-09-04  AAPL_open  120.070000
4  2020-09-08  AAPL_open  113.949997

In [87]:
var_names = close_df.columns.to_list()
var_names.remove('date')
close_stack = close_df.reset_index().melt( id_vars = ['date'], value_vars = var_names )
close_stack.head()

date    variable       value
0  2020-09-01  AAPL_close  134.179993
1  2020-09-02  AAPL_close  131.399994
2  2020-09-03  AAPL_close  120.879997
3  2020-09-04  AAPL_close  120.959999
4  2020-09-08  AAPL_close  112.820000

In [88]:
var_names = low_df.columns.to_list()
var_names.remove('date')
low_stack = low_df.reset_index().melt( id_vars = ['date'], value_vars = var_names )
low_stack.head()

date  variable       value
0  2020-09-01  AAPL_low  130.529999
1  2020-09-02  AAPL_low  127.000000
2  2020-09-03  AAPL_low  120.500000
3  2020-09-04  AAPL_low  110.889999
4  2020-09-08  AAPL_low  112.680000

In [89]:
var_names = high_df.columns.to_list()
var_names.remove('date')
high_stack = high_df.reset_index().melt( id_vars = ['date'], value_vars = var_names )
high_stack.head()

date   variable       value
0  2020-09-01  AAPL_high  134.800003
1  2020-09-02  AAPL_high  137.979996
2  2020-09-03  AAPL_high  128.839996
3  2020-09-04  AAPL_high  123.699997
4  2020-09-08  AAPL_high  118.989998

#### Split the new column into two

The new column represents two variables - stock information and price information. Split this information into two new variables, symbol and price_type, and drop the variable column.

In [90]:
open_stack[['symbol','price_type']] = open_stack.variable.str.split( '_', expand=True)
open_stack.drop(['variable'], axis=1, inplace=True)
open_stack.head()

date       value symbol price_type
0  2020-09-01  132.759995   AAPL       open
1  2020-09-02  137.589996   AAPL       open
2  2020-09-03  126.910004   AAPL       open
3  2020-09-04  120.070000   AAPL       open
4  2020-09-08  113.949997   AAPL       open

In [91]:
close_stack[['symbol','price_type']] = close_stack.variable.str.split( '_', expand=True)
close_stack.drop(['variable'], axis=1, inplace=True)
close_stack.head()

date       value symbol price_type
0  2020-09-01  134.179993   AAPL      close
1  2020-09-02  131.399994   AAPL      close
2  2020-09-03  120.879997   AAPL      close
3  2020-09-04  120.959999   AAPL      close
4  2020-09-08  112.820000   AAPL      close

In [92]:
low_stack[['symbol','price_type']] = low_stack.variable.str.split( '_', expand=True)
low_stack.drop(['variable'], axis=1, inplace=True)
low_stack.head()

date       value symbol price_type
0  2020-09-01  130.529999   AAPL        low
1  2020-09-02  127.000000   AAPL        low
2  2020-09-03  120.500000   AAPL        low
3  2020-09-04  110.889999   AAPL        low
4  2020-09-08  112.680000   AAPL        low

In [93]:
high_stack[['symbol','price_type']] = high_stack.variable.str.split( '_', expand=True)
high_stack.drop(['variable'], axis=1, inplace=True)
high_stack.head()

date       value symbol price_type
0  2020-09-01  134.800003   AAPL       high
1  2020-09-02  137.979996   AAPL       high
2  2020-09-03  128.839996   AAPL       high
3  2020-09-04  123.699997   AAPL       high
4  2020-09-08  118.989998   AAPL       high

#### Combine datasets

Concatenate the stock prices datasets into one. Check unique values of symbol in both datasets  and merge with the info_df dataset on symbol.

In [94]:
prices_df = pd.concat( [open_stack, close_stack, low_stack, high_stack], ignore_index=True)
prices_df

date       value symbol price_type
0     2020-09-01  132.759995   AAPL       open
1     2020-09-02  137.589996   AAPL       open
2     2020-09-03  126.910004   AAPL       open
3     2020-09-04  120.070000   AAPL       open
4     2020-09-08  113.949997   AAPL       open
...          ...         ...    ...        ...
2395  2021-01-19   57.580002     VZ       high
2396  2021-01-20   57.389999     VZ       high
2397  2021-01-21   57.509998     VZ       high
2398  2021-01-22   57.689999     VZ       high
2399  2021-01-25   58.619999     VZ       high

[2400 rows x 4 columns]

In [95]:
prices_df.symbol.value_counts()

AAPL    400
VZ      400
GME     400
KO      400
TSLA    400
AMZN    400
Name: symbol, dtype: int64

In [96]:
info_df.symbol.value_counts()

TSLA    1
VZ      1
AMZN    1
KO      1
GME     1
AAPL    1
Name: symbol, dtype: int64

In [98]:
stock_prices = pd.merge(
    prices_df,
    info_df,
    how="left",
    on="symbol",
)

In [ ]:
stock_prices.to_csv('stocks_tidy.csv')

In [99]:
stock_prices.drop(['symbol'], axis=1, inplace=True)
stock_prices

date       value price_type                      company  \
0     2020-09-01  132.759995       open                   Apple Inc.   
1     2020-09-02  137.589996       open                   Apple Inc.   
2     2020-09-03  126.910004       open                   Apple Inc.   
3     2020-09-04  120.070000       open                   Apple Inc.   
4     2020-09-08  113.949997       open                   Apple Inc.   
...          ...         ...        ...                          ...   
2395  2021-01-19   57.580002       high  Verizon Communications Inc.   
2396  2021-01-20   57.389999       high  Verizon Communications Inc.   
2397  2021-01-21   57.509998       high  Verizon Communications Inc.   
2398  2021-01-22   57.689999       high  Verizon Communications Inc.   
2399  2021-01-25   58.619999       high  Verizon Communications Inc.   

                      sector exchange  
0     Information Technology   NASDAQ  
1     Information Technology   NASDAQ  
2     Information Technology   NASDAQ  
3     Information Technology   NASDAQ  
4     Information Technology   NASDAQ  
...                      ...      ...  
2395  Communication Services     NYSE  
2396  Communication Services     NYSE  
2397  Communication Services     NYSE  
2398  Communication Services     NYSE  
2399  Communication Services     NYSE  

[2400 rows x 6 columns]

#### Discussion on the tidy stock market data

There are six companies represented in the final, tidy, stock prices dataset, and this is consistent with the original separated stock price data files, which had these companies represented in the header (six company-price_type variables in each). The stock_prices dataset has seven columns and 2400 rows with 24 rows for each day.

In [100]:
stock_prices.company.value_counts()

Verizon Communications Inc.    400
Coca-Cola Company              400
GameStop Corp.                 400
Amazon.com Inc.                400
Tesla Inc                      400
Apple Inc.                     400
Name: company, dtype: int64

In [101]:
stock_prices.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2400 entries, 0 to 2399
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        2400 non-null   object 
 1   value       2400 non-null   float64
 2   price_type  2400 non-null   object 
 3   company     2400 non-null   object 
 4   sector      2400 non-null   object 
 5   exchange    2400 non-null   object 
dtypes: float64(1), object(5)
memory usage: 131.2+ KB


In [102]:
total_cases = stock_prices.groupby(['date']).\
aggregate(total_cases = ('value', 'count')).\
reset_index().sort_values(by=["date"])

In [60]:
check_totals = total_cases.loc[total_cases['total_cases'] != 24]
check_totals

Empty DataFrame
Columns: [date, total_cases]
Index: []

## Problem 04: Group and Aggregate the Stock Market Data

#### How many unique days are associated with each MONTH in the data set?

Convert date to datetime and extract the month, day, and year into three new columns. Using that dataset, group by month and year (although the months alone are unique) and count the number of unique days.

In [61]:
stock_prices.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2400 entries, 0 to 2399
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        2400 non-null   object 
 1   value       2400 non-null   float64
 2   price_type  2400 non-null   object 
 3   company     2400 non-null   object 
 4   sector      2400 non-null   object 
 5   exchange    2400 non-null   object 
dtypes: float64(1), object(5)
memory usage: 195.8+ KB


In [62]:
stock_copy = stock_prices.copy()
stock_copy.head()

date       value price_type     company                  sector  \
0  2020-09-01  132.759995       open  Apple Inc.  Information Technology   
1  2020-09-02  137.589996       open  Apple Inc.  Information Technology   
2  2020-09-03  126.910004       open  Apple Inc.  Information Technology   
3  2020-09-04  120.070000       open  Apple Inc.  Information Technology   
4  2020-09-08  113.949997       open  Apple Inc.  Information Technology   

  exchange  
0   NASDAQ  
1   NASDAQ  
2   NASDAQ  
3   NASDAQ  
4   NASDAQ

In [63]:
stock_copy['date_dt'] = pd.to_datetime( stock_copy.date )
stock_copy['year'] = stock_copy.date_dt.dt.year
stock_copy['month'] = stock_copy.date_dt.dt.month
stock_copy['day'] = stock_copy.date_dt.dt.day
stock_copy.drop(['date'], axis=1, inplace=True)
stock_copy.head()

value price_type     company                  sector exchange  \
0  132.759995       open  Apple Inc.  Information Technology   NASDAQ   
1  137.589996       open  Apple Inc.  Information Technology   NASDAQ   
2  126.910004       open  Apple Inc.  Information Technology   NASDAQ   
3  120.070000       open  Apple Inc.  Information Technology   NASDAQ   
4  113.949997       open  Apple Inc.  Information Technology   NASDAQ   

     date_dt  year  month  day  
0 2020-09-01  2020      9    1  
1 2020-09-02  2020      9    2  
2 2020-09-03  2020      9    3  
3 2020-09-04  2020      9    4  
4 2020-09-08  2020      9    8

In [64]:
stock_copy.groupby(['month', 'year']).\
aggregate(unique_days = ('day', 'nunique')).\
reset_index()

month  year  unique_days
0      1  2021           15
1      9  2020           21
2     10  2020           22
3     11  2020           20
4     12  2020           22

#### Calculate the difference between the daily close and daily open prices for each stock.

Using the dataset with datetime values, pivot the table on price_type so daily open and close values are column variables and create two new columns for the difference and the absolute value difference between daily close-to-open prices.

In [65]:
stock_open_close = stock_copy.loc[ stock_copy.price_type.isin(['open','close']), : ].copy()
stock_open_close.head()

value price_type     company                  sector exchange  \
0  132.759995       open  Apple Inc.  Information Technology   NASDAQ   
1  137.589996       open  Apple Inc.  Information Technology   NASDAQ   
2  126.910004       open  Apple Inc.  Information Technology   NASDAQ   
3  120.070000       open  Apple Inc.  Information Technology   NASDAQ   
4  113.949997       open  Apple Inc.  Information Technology   NASDAQ   

     date_dt  year  month  day  
0 2020-09-01  2020      9    1  
1 2020-09-02  2020      9    2  
2 2020-09-03  2020      9    3  
3 2020-09-04  2020      9    4  
4 2020-09-08  2020      9    8

In [66]:
stock_pivot = stock_open_close.pivot(index=['date_dt', 'year', 'month', 'day', 'company', 'sector', 'exchange'], columns='price_type', values='value').\
reset_index()
stock_pivot.head()

price_type    date_dt  year  month  day            company  \
0          2020-09-01  2020      9    1    Amazon.com Inc.   
1          2020-09-01  2020      9    1         Apple Inc.   
2          2020-09-01  2020      9    1  Coca-Cola Company   
3          2020-09-01  2020      9    1     GameStop Corp.   
4          2020-09-01  2020      9    1          Tesla Inc   

price_type                  sector exchange        close         open  
0           Consumer Discretionary   NASDAQ  3499.120117  3489.580078  
1           Information Technology   NASDAQ   134.179993   132.759995  
2                 Consumer Staples     NYSE    49.119999    49.310001  
3           Consumer Discretionary     NYSE     7.650000     7.300000  
4           Consumer Discretionary   NASDAQ   475.049988   502.140015

In [67]:
stock_pivot['difference'] = stock_pivot['close'] - stock_pivot['open']
stock_pivot.head()

price_type    date_dt  year  month  day            company  \
0          2020-09-01  2020      9    1    Amazon.com Inc.   
1          2020-09-01  2020      9    1         Apple Inc.   
2          2020-09-01  2020      9    1  Coca-Cola Company   
3          2020-09-01  2020      9    1     GameStop Corp.   
4          2020-09-01  2020      9    1          Tesla Inc   

price_type                  sector exchange        close         open  \
0           Consumer Discretionary   NASDAQ  3499.120117  3489.580078   
1           Information Technology   NASDAQ   134.179993   132.759995   
2                 Consumer Staples     NYSE    49.119999    49.310001   
3           Consumer Discretionary     NYSE     7.650000     7.300000   
4           Consumer Discretionary   NASDAQ   475.049988   502.140015   

price_type  difference  
0             9.540039  
1             1.419998  
2            -0.190002  
3             0.350000  
4           -27.090027

In [68]:
stock_pivot['abs_difference'] = abs(stock_pivot['difference'])
stock_pivot['diff_direction'] = np.where(stock_pivot['difference'] >= 0,'gain','loss')
stock_pivot.head()

price_type    date_dt  year  month  day            company  \
0          2020-09-01  2020      9    1    Amazon.com Inc.   
1          2020-09-01  2020      9    1         Apple Inc.   
2          2020-09-01  2020      9    1  Coca-Cola Company   
3          2020-09-01  2020      9    1     GameStop Corp.   
4          2020-09-01  2020      9    1          Tesla Inc   

price_type                  sector exchange        close         open  \
0           Consumer Discretionary   NASDAQ  3499.120117  3489.580078   
1           Information Technology   NASDAQ   134.179993   132.759995   
2                 Consumer Staples     NYSE    49.119999    49.310001   
3           Consumer Discretionary     NYSE     7.650000     7.300000   
4           Consumer Discretionary   NASDAQ   475.049988   502.140015   

price_type  difference  abs_difference diff_direction  
0             9.540039        9.540039           gain  
1             1.419998        1.419998           gain  
2            -0.190002        0.190002           loss  
3             0.350000        0.350000           gain  
4           -27.090027       27.090027           loss

#### What is the maximum daily close-to-open difference for each company?

Using the pivoted dataset with the calculated differences, group by company and obtain the maximum close-to-open difference per company. The table below shows the maximum positive difference, the maximum negative difference, and the maximum absolute value difference with a column that indicates the direction of the absolute value difference as gain or loss.

In [69]:
max_difference_co = stock_pivot.groupby(['company']).\
aggregate(max_positive_diff = ('difference', 'max'),
          max_negative_diff = ('difference', 'min'),
          max_absolute_diff = ('abs_difference', 'max')).\
reset_index()

In [70]:
max_difference_co['max_positive_diff'] = max_difference_co['max_positive_diff'].apply(lambda x: "${:.2f}".format((x)))
max_difference_co['max_negative_diff'] = max_difference_co['max_negative_diff'].apply(lambda x: "${:.2f}".format((x)))
max_difference_co['max_absolute_diff'] = max_difference_co['max_absolute_diff'].apply(lambda x: "${:.2f}".format((x)))
max_difference_co['max_abs_direction'] = np.where(max_difference_co['max_absolute_diff'] == max_difference_co['max_positive_diff'],'gain','loss')
max_difference_co

company max_positive_diff max_negative_diff  \
0              Amazon.com Inc.            $95.15          $-132.11   
1                   Apple Inc.             $5.54            $-6.87   
2            Coca-Cola Company             $1.83            $-1.52   
3               GameStop Corp.            $22.42           $-19.94   
4                    Tesla Inc            $52.70           $-49.21   
5  Verizon Communications Inc.             $1.38            $-1.03   

  max_absolute_diff max_abs_direction  
0           $132.11              loss  
1             $6.87              loss  
2             $1.83              gain  
3            $22.42              gain  
4            $52.70              gain  
5             $1.38              gain

#### What is the maximum daily close-to-open difference for each company in each MONTH?

Using the pivoted dataset with the calculated differences, group by company, month, and year and obtain the maximum close-to-open stock price per company per month. The table below shows the maximum positive difference, the maximum negative difference, and the maximum absolute value difference with a column that indicates the direction of the absolute value difference as gain or loss.

In [71]:
max_difference_co_mo = stock_pivot.groupby(['company', 'month', 'year']).\
aggregate(max_positive_diff = ('difference', 'max'),
          max_negative_diff = ('difference', 'min'),
          max_absolute_diff = ('abs_difference', 'max')).\
reset_index()

In [72]:
max_difference_co_mo['max_positive_diff'] = max_difference_co_mo['max_positive_diff'].apply(lambda x: "${:.2f}".format((x)))
max_difference_co_mo['max_negative_diff'] = max_difference_co_mo['max_negative_diff'].apply(lambda x: "${:.2f}".format((x)))
max_difference_co_mo['max_absolute_diff'] = max_difference_co_mo['max_absolute_diff'].apply(lambda x: "${:.2f}".format((x)))
max_difference_co_mo['max_abs_direction'] = np.where(max_difference_co_mo['max_absolute_diff'] == max_difference_co_mo['max_positive_diff'],'gain','loss')
max_difference_co_mo

company  month  year max_positive_diff  \
0               Amazon.com Inc.      1  2021            $81.39   
1               Amazon.com Inc.      9  2020            $95.15   
2               Amazon.com Inc.     10  2020            $92.99   
3               Amazon.com Inc.     11  2020            $81.17   
4               Amazon.com Inc.     12  2020            $89.96   
5                    Apple Inc.      1  2021             $3.37   
6                    Apple Inc.      9  2020             $5.54   
7                    Apple Inc.     10  2020             $4.34   
8                    Apple Inc.     11  2020             $2.30   
9                    Apple Inc.     12  2020             $3.54   
10            Coca-Cola Company      1  2021             $1.05   
11            Coca-Cola Company      9  2020             $1.83   
12            Coca-Cola Company     10  2020             $0.75   
13            Coca-Cola Company     11  2020             $1.41   
14            Coca-Cola Company     12  2020             $1.07   
15               GameStop Corp.      1  2021            $22.42   
16               GameStop Corp.      9  2020             $1.65   
17               GameStop Corp.     10  2020             $3.95   
18               GameStop Corp.     11  2020             $1.15   
19               GameStop Corp.     12  2020             $3.24   
20                    Tesla Inc      1  2021            $38.41   
21                    Tesla Inc      9  2020            $38.67   
22                    Tesla Inc     10  2020            $11.52   
23                    Tesla Inc     11  2020            $38.29   
24                    Tesla Inc     12  2020            $52.70   
25  Verizon Communications Inc.      1  2021             $0.95   
26  Verizon Communications Inc.      9  2020             $1.38   
27  Verizon Communications Inc.     10  2020             $0.60   
28  Verizon Communications Inc.     11  2020             $0.82   
29  Verizon Communications Inc.     12  2020             $0.69   

   max_negative_diff max_absolute_diff max_abs_direction  
0            $-83.37            $83.37              loss  
1           $-132.11           $132.11              loss  
2           $-121.60           $121.60              loss  
3            $-87.29            $87.29              loss  
4            $-63.69            $89.96              gain  
5             $-4.11             $4.11              loss  
6             $-6.87             $6.87              loss  
7             $-4.17             $4.34              gain  
8             $-4.18             $4.18              loss  
9             $-3.18             $3.54              gain  
10            $-1.51             $1.51              loss  
11            $-1.52             $1.83              gain  
12            $-1.04             $1.04              loss  
13            $-1.14             $1.41              gain  
14            $-0.77             $1.07              gain  
15           $-19.94            $22.42              gain  
16            $-0.61             $1.65              gain  
17            $-1.48             $3.95              gain  
18            $-1.61             $1.61              loss  
19            $-1.44             $3.24              gain  
20           $-38.21            $38.41              gain  
21           $-31.62            $38.67              gain  
22           $-18.86            $18.86              loss  
23           $-34.61            $38.29              gain  
24           $-49.21            $52.70              gain  
25            $-0.43             $0.95              gain  
26            $-0.92             $1.38              gain  
27            $-1.03             $1.03              loss  
28            $-0.81             $0.82              gain  
29            $-0.75             $0.75              loss

#### What is the average daily close-to-open difference for each company in each MONTH?

Using the pivoted dataset with the calculated differences, group by company, month, and year and obtain the average close-to-open difference per company per month.

In [73]:
avg_diff_co_mo = stock_pivot.groupby(['company', 'month', 'year']).\
aggregate(mean_difference = ('difference', 'mean')).\
reset_index()

In [74]:
avg_diff_co_mo['mean_difference'] = avg_diff_co_mo['mean_difference'].apply(lambda x: "${:.2f}".format((x)))
avg_diff_co_mo

company  month  year mean_difference
0               Amazon.com Inc.      1  2021          $-1.53
1               Amazon.com Inc.      9  2020         $-19.79
2               Amazon.com Inc.     10  2020         $-10.76
3               Amazon.com Inc.     11  2020          $-5.52
4               Amazon.com Inc.     12  2020          $-2.18
5                    Apple Inc.      1  2021           $0.46
6                    Apple Inc.      9  2020          $-0.95
7                    Apple Inc.     10  2020          $-0.28
8                    Apple Inc.     11  2020           $0.08
9                    Apple Inc.     12  2020           $0.14
10            Coca-Cola Company      1  2021          $-0.24
11            Coca-Cola Company      9  2020          $-0.04
12            Coca-Cola Company     10  2020          $-0.03
13            Coca-Cola Company     11  2020          $-0.12
14            Coca-Cola Company     12  2020           $0.13
15               GameStop Corp.      1  2021           $0.97
16               GameStop Corp.      9  2020           $0.06
17               GameStop Corp.     10  2020           $0.03
18               GameStop Corp.     11  2020           $0.02
19               GameStop Corp.     12  2020           $0.05
20                    Tesla Inc      1  2021           $4.39
21                    Tesla Inc      9  2020          $-0.79
22                    Tesla Inc     10  2020          $-3.28
23                    Tesla Inc     11  2020           $1.34
24                    Tesla Inc     12  2020           $7.31
25  Verizon Communications Inc.      1  2021           $0.09
26  Verizon Communications Inc.      9  2020          $-0.01
27  Verizon Communications Inc.     10  2020          $-0.08
28  Verizon Communications Inc.     11  2020           $0.02
29  Verizon Communications Inc.     12  2020          $-0.06

#### Which day had the maximum daily high price for each company in each MONTH?

Using a dataset with only the 'high' price_type observations, group by company, month, and year and obtain the maximum high stock price per company per month. 

In [75]:
stock_high = stock_copy.loc[ stock_copy.price_type.isin(['high']), : ].copy()
stock_high.head()

value price_type     company                  sector exchange  \
1800  134.800003       high  Apple Inc.  Information Technology   NASDAQ   
1801  137.979996       high  Apple Inc.  Information Technology   NASDAQ   
1802  128.839996       high  Apple Inc.  Information Technology   NASDAQ   
1803  123.699997       high  Apple Inc.  Information Technology   NASDAQ   
1804  118.989998       high  Apple Inc.  Information Technology   NASDAQ   

        date_dt  year  month  day  
1800 2020-09-01  2020      9    1  
1801 2020-09-02  2020      9    2  
1802 2020-09-03  2020      9    3  
1803 2020-09-04  2020      9    4  
1804 2020-09-08  2020      9    8

In [76]:
max_high = stock_high.groupby(['company', 'month', 'year']).\
aggregate(max_high = ('value', 'max')).\
reset_index()

In [77]:
max_high['max_high'] = max_high['max_high'].apply(lambda x: "${:.2f}".format((x)))
max_high

company  month  year  max_high
0               Amazon.com Inc.      1  2021  $3363.89
1               Amazon.com Inc.      9  2020  $3552.25
2               Amazon.com Inc.     10  2020  $3496.24
3               Amazon.com Inc.     11  2020  $3366.80
4               Amazon.com Inc.     12  2020  $3350.65
5                    Apple Inc.      1  2021   $145.09
6                    Apple Inc.      9  2020   $137.98
7                    Apple Inc.     10  2020   $125.39
8                    Apple Inc.     11  2020   $121.99
9                    Apple Inc.     12  2020   $138.79
10            Coca-Cola Company      1  2021    $54.63
11            Coca-Cola Company      9  2020    $52.14
12            Coca-Cola Company     10  2020    $51.53
13            Coca-Cola Company     11  2020    $54.62
14            Coca-Cola Company     12  2020    $54.93
15               GameStop Corp.      1  2021   $159.18
16               GameStop Corp.      9  2020    $11.17
17               GameStop Corp.     10  2020    $15.87
18               GameStop Corp.     11  2020    $19.42
19               GameStop Corp.     12  2020    $22.35
20                    Tesla Inc      1  2021   $900.40
21                    Tesla Inc      9  2020   $502.49
22                    Tesla Inc     10  2020   $465.90
23                    Tesla Inc     11  2020   $607.80
24                    Tesla Inc     12  2020   $718.72
25  Verizon Communications Inc.      1  2021    $59.54
26  Verizon Communications Inc.      9  2020    $61.50
27  Verizon Communications Inc.     10  2020    $59.97
28  Verizon Communications Inc.     11  2020    $61.46
29  Verizon Communications Inc.     12  2020    $61.95